### ISSUE TITLE
Where should new data structures placed

### ISSUE THREAD

**Issue by Rapiz1** [chapel-lang/chapel#15950]
I have been working on the GSoC DS project for a while. I feel it's needed to discuss the location of the structures proposed by this project.
## Heap
I think it should be in package modules. It's commonly used but not that essential. But I'm not strongly against adding to standard modules.
## Vector
In https://github.com/chapel-lang/chapel/issues/15913#issuecomment-648274188 @bradcray commented
> Also, as I've said before, I don't think this complete architecture / refactoring needs to be completed before any vector-oriented list code is merged to master. It would be fine with me if Vector showed up as a separate module (by some name) as an initial step to lock some code in and then we worked on refactoring the various implementations to all live under List.

So I suppose I could first open a PR to merge it into package modules, and then open a PR to refactor and move it into standard modules. 

## OSet/OMap
*The names haven't been settled yet.*
I think this is fundamental enough to get placed in standard modules. For example, C++ calls the ordered map and set as `std::map` and `std::set`

## UnrolledLinkedList
I expect the same as Vector. It could be an underlying implementation for `LinkedList`. So firstly package modules and then move it into standard modules.

**Comment by Rapiz1** [comment]
FYI: @krishnadey30 @e-kayrakli @cassella

**Comment by bradcray** [comment]
My opinions:

I'd probably put heap into standard modules.  It seems sufficiently like a sibling to the other collections to belong there.

Putting Vector into packages first and then into standard once it sits below List sounds good to me.

OSet/OMap: I'm worried I'm becoming predictable, but...  has anyone kicked around the idea of having an `ordered=true`/`ordered=false` parameter on the normal `set`/`map` types?  Are the interfaces sufficiently similar that this could work?

For UnrolledLinkedList, packages sounds like a good starting point.  Rather than putting it under LinkedList, I'd probably make it yet another option under List (though... does the user specify the unrolling factor?)

**Comment by Rapiz1** [comment]
> OSet/OMap: I'm worried I'm becoming predictable, but... has anyone kicked around the idea of having an `ordered=true`/`ordered=false` parameter on the normal `set`/`map` types? Are the interfaces sufficiently similar that this could work?

I don't think that's a good idea.

1. OSet/OMap is designed to utilize Treap, SkipList, or some ordered structures(e.g. Red-Black Tree) we may add later, while Map utilizes HashTable. I think the differences like O(1) vs O(lgN), ordered structures vs unordered structures, are significant enough to split up Set and OSet. In addition, possibly we have far more ordered structures(various trees) than unordered structures. It makes more sense to group them together.

1. Also, OSet supports more operations than Set and that's not something like Vector has just one more procedure `requestCapacity`.You can check out [here](https://github.com/chapel-lang/chapel/issues/15921). The use cases are also slightly different. With Set, users care about whether the element exists. With OSet, users care more about the relationship between elements. Like, what's the predecessor of `x`?
The use case of OMap is very similar to Map. But due to 1. and the consistency, I think it shouldn't be combined with Map.

> For UnrolledLinkedList, packages sounds like a good starting point. Rather than putting it under LinkedList, I'd probably make it yet another option under List (though... does the user specify the unrolling factor?)

Yes. Users can specify the unrolling factor and that's an important feature that enables users to use their cache wisely. 

Sounds wired to have it in List. If users specify one element in each block, then UnrolledLinkedList just becomes LinkedList. They share much similarity. Moreover, as a user, I expect List supports very quick random access. That's true for both Vector and current List, but not for (Unrolled)LinkedList. 

Another personal opinion is that (Unrolled)LinkedList could be far more powerful if supporting slicing and linking nodes. I feel that's the main reason I usually use linked structures in programming. Putting UnrolledLinkedList limits that possibility. However, nobody is proposing that, and [that's out of scope for this project](https://github.com/chapel-lang/chapel/issues/15669#issuecomment-632564639). So just my personal opinions.

**Comment by bradcray** [comment]
I buy the argument that if the interfaces are significantly different there's less of a reason to unify them.  I think we could still make it work if someone really wanted to advocate for this, though (e.g., by having the outer `set` datatype forward any unrecognized method calls to the implementing record), but I won't fight for it.

I'm less convinced by differences in asymptotic or practical timings between operations being a reason not to unify, though, and actually consider it a strong reason to unify.  In selecting a datatype, I primarily care about the operations it supports, secondarily about the implementation and its effect on the performance of those operations.  It seems convenient to me to choose between these implementations via an argument to list rather than needing to change a type name everywhere I mention it.  Even implementations that we're considering unifying have dramatically different implementation costs (e.g., inserting into a vector-style list can sometimes be O(1), other times O(n); inserting into a multi-vector list is always O(1) though sometimes it's more expensive than others if we need to allocate a new buffer).

I had imagined LinkedList also being an implementation option under `list` which is why I was suggesting putting UnrolledLinkedList there as well.  Can you say more about what you mean w.r.t. supporting slicing and linking nodes, though?  I'm not sure I understand, though I have a guess.

**Comment by Rapiz1** [comment]
> Can you say more about what you mean w.r.t. supporting slicing and linking nodes, though? I'm not sure I understand, though I have a guess.

Provide users with reference to nodes somehow and allow cutting nodes off from the (Unrolled)LinkedList and then link them somewhere. Users can modify the order of a sequence of elements efficiently.

**Comment by cassella** [comment]
Another option for the initial commit of `Vector` is to make it a test module.  I thought there was a practice of putting modules under `test/` while they're under development -- in part so users don't discover them and start relying on the interface -- but I can't find an example right now.

**Comment by Rapiz1** [comment]
I also want to add that LinkedList should have O(1) insertion if somehow we offer users references to nodes. If it's simply under List then `insert(idx, element)` will cost O(N) to locate the node and then insert. That's a big drawback for LinkedList. And any improvement in this direction means a non-trivial change to List API.

**Comment by e-kayrakli** [comment]
I am a bit worried that this issue brings two discussion points together. Where should the data structures go is a bit different than what they should be named and/or whether they should be incorporated into the existing interfaces. I think we already have an issue specific for OSet/OMap as mentioned here under the issue https://github.com/chapel-lang/chapel/issues/15921. I'll put some comments regarding names there, citing some of the discussion here.

- Heap: Is an important structure to me, and probably in the same class of data structures we have in standard modules. So, I think standard modules is the right place for it
- Vector: See https://github.com/chapel-lang/chapel/issues/15913. As we want it to be an alternative `List`, it'll be in standard modules eventually. I don't have a much of an opinion about the path it takes to get there, as long as it happens before the release, which is months away.
- OSet/OMap: Don't know how I feel about these. Fundamentally I see "map" and "set" as unordered data structures, and see these as a variation of those and as such not that "important". It is a good argument that `std::map` and `std::set` are ordered, but I don't mind breaking precedence there personally. This also depends on whether we want these to be just implementations of set and map, in which case, they'll follow the path of Vector in this list.
- UnrolledLinkedList: The argument of potentially faster insertion is not enough for me to making this a separate structure. I think we should try hard to make this an implementation of List. And naturally, if we do that it'll also follow the path of Vector.

**Comment by Rapiz1** [comment]
As @e-kayrakli said, there are two questions about each structure. I suggest we put the design discussion in their respective issues, including the discussion about [UnrolledLinkedList](https://github.com/chapel-lang/chapel/issues/15669).
It makes sense that we discuss Vector and Heap here because they are relatively settled.

**Comment by Rapiz1** [comment]
> Another option for the initial commit of `Vector` is to make it a test module. I thought there was a practice of putting modules under `test/` while they're under development -- in part so users don't discover them and start relying on the interface -- but I can't find an example right now.

>  *(quote from gitter)* There's not a particular directory, but in some cases the module is checked in for testing in the test/library/x/Y directory alongside the tests for the module. For example, there's test/library/packages/Itertools/, which has a few directories of tests, but also Itertools.chpl that's the proposed module, and Itertools.notest because the proposed module isn't a test.
Though on reflection, I think that treatment is particularly for when a module is checked in before a release but won't be finished before that release. If the module will be checked in and completed in a single release cycle, it can probably go straight into modules/. (At least, that's my understanding now.) 

Given that nobody proposes names for the standalone Vector, so merging into `packages` is not that feasible. I think put `Vector` into `test` is not a bad idea. Lacking a name also means uncompleted to some extent. As an initial step, `Vector` should be reviewed and merged into `test`, lessening some burden when reviewing the refactor of all the stuff about List. Then at some point, List directly shows up with a new interface. This way also means less breaking change (adding `Vector` and moving `Vector`).

I can't come up with something better if `Vector` doesn't have a new name.

**Comment by bradcray** [comment]
I think adding a Vector module to the test system would be fine.  As mentioned on chat, I'd put it in a directory like `test/library/draft/Vector/` or `test/library/dev/Vector/` or `test/library/inProgress/Vector/`.

**Comment by e-kayrakli** [comment]
We have added

- Heap: As a standard module
- OrderedSet/OrderedMap/UnrolledLinkedList: As a package module
- Vector: As a draft module. (We'll need to figure out how `vector` should coexist with `list)
